test/t8_schemes/t8_gtest_element_ref_coords.cxx

```diff
@@ -0,0 +1,184 @@
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with t8code; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*/
+
+/** \file t8_gtest_element_ref_coords.cxx
+* Provide tests to check the functionality of the computation of
+* element reference coordinates and element centroids.
+*/
+
+#include <gtest/gtest.h>
+#include <t8_eclass.h>
+#include <t8_vec.h>
+#include <t8_element_cxx.hxx>
+#include <t8_schemes/t8_default/t8_default_cxx.hxx>
+#include <t8_forest/t8_forest.h>
+#include <t8_cmesh/t8_cmesh_examples.h>
+
+#if T8_ENABLE_LESS_TESTS
+#define MAXLEVEL 3
+#else
+#define MAXLEVEL 4
+#endif
+
+void
+t8_element_centroid_by_vertex_coords (t8_forest_t forest, t8_eclass_scheme_c *ts, t8_locidx_t ltreeid,
+                                      const t8_element_t *element, double *coordinates)
+{
+  double vertex_ref_coords[3], vertex_out_coords[3];
+  t8_gloidx_t gtreeid = t8_forest_global_tree_id (forest, ltreeid);
+  t8_cmesh_t cmesh = t8_forest_get_cmesh (forest);
+
+  /* Initialize the centroid with (0, 0, 0). */
+  memset (coordinates, 0, 3 * sizeof (double));
+  /* Get the number of corners of the element. */
+  const int num_vertices = ts->t8_element_num_corners (element);
+  for (int i_vertex = 0; i_vertex < num_vertices; i_vertex++) {
+    /* For each corner, add its coordinates to the centroids coordinates. */
+
+    /* Compute the vertex coordinates inside [0,1]^dim reference cube. */
+    ts->t8_element_vertex_reference_coords (element, i_vertex, vertex_ref_coords);
+    /* Evaluate the geometry */
+    t8_geometry_evaluate (cmesh, gtreeid, vertex_ref_coords, 1, vertex_out_coords);
+    /* coordinates = coordinates + vertex_coords */
+    t8_vec_axpy (vertex_out_coords, coordinates, 1);
+  }
+  /* Divide each coordinate by num_vertices */
+  t8_vec_ax (coordinates, 1. / num_vertices);
+}
+
+void
+t8_get_batch_coords_for_element_type (const t8_element_shape_t shape, double *batch_coords)
+{
+  const int num_vertices = t8_eclass_num_vertices[shape];
+  const int elem_dim = t8_eclass_to_dimension[shape];
+  for (int i_vertex = 0; i_vertex < num_vertices; ++i_vertex) {
+    for (int dim = 0; dim < elem_dim; ++dim) {
+      batch_coords[i_vertex * elem_dim + dim] = t8_element_corner_ref_coords[shape][i_vertex][dim];
+    }
+  }
+}
+
+template <typename T>
+void
+t8_write_message_by_dim (const char *message, const T *array, const int dim)
+{
+  std::ostringstream buffer;
+  buffer.precision (6);
+  buffer.setf (std::ios::fixed);
+  buffer << message;
+  for (int i_dim = 0; i_dim < dim; ++i_dim) {
+    buffer << " " << array[i_dim];
+  }
+  t8_debugf ("%s\n", buffer.str ().c_str ());
+}
+
+/* To save an iterate, we will test the ref coords and centroid right in the adapt callback itself. 
+ * Refines as long as the MAXLEVEL is not reached. */
```

Comment:
I think this optimization is a little bit to much. If we ever encounter a bug in the ref_coords computation again it will make debugging more complicated. The purpose of the tests are not to run the most optimal way, but to ensure all functionalities of t8code and to make our lifes easier, when we hunt for bugs. 
Seperating adaptation and testing will make it easier.

Reply:
Yes, but since the adapt callback just returns 1, there will be no bugs in the adaptation itself. I see no advantage in using an adapt callback which just returns 1 and another iterate, which would itself add more complexity.

Reply:
We  just don't know the error that will make it crash yet. 
It easiert to debug a function that does only one thing, than a function that does two things. 
What kind of complexity do you mean? The runtime complexity, or the ccomplexity to understand the code?

Reply:
If your adaptation returns 1, you actually compute a uniform forest on each level. For that purpose it is better to use t8_forest_new_uniform (as you don't have any data that is passed from the original to the refined forest).

Reply:
Changed that :)